Applying PDA overhaul Upstream

TungstenOxide opened the pull request:
🆑
tweak: Added a few html snippets to change the font of the PDA
tweak: Added a few html snippets to remove bullet points and underlines from PDA menus
tweak: Made all PDA icons transparent so that the color change would actually affect them
add: some PDA icons
add: option to cycle through three different fonts for the PDA
add: option to change color to literally whatever you want
/🆑

[why]:
Cleaning up the PDA and making it look a little nicer and more authentic. So, A E S T H E T I C . Also, added some customization options. HMU if you have a suggestion for a new font.

Here's a GIF of it in action : https://imgur.com/a/yz5P6

TungstenOxide committed:
Applying PDA overhaul Upstream

Yeet?

lordpidey commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -307,6 +316,24 @@ GLOBAL_LIST_EMPTY(PDAs)
+				if (font_index == 3)
+					font_index = 0
+				else
+					font_index++
```
Replace this whole if statement with
```
font_index++
font_index = font_index%4
```

TungstenOxide replied:
@lordpidey Did I do it right?

TungstenOxide replied:
I think that might have broken it.

lordpidey replied:
No, you need to replace lines 320 through 323 with what I posted.  You also took out 319.

TungstenOxide replied:
Oh whoops

TungstenOxide replied:
Okay it should work now.

TungstenOxide replied:
@lordpidey I did the thing

lordpidey commented:
Looks like a good addition, thank you for backporting this.

TungstenOxide committed:
Used lordpidey's change

Like this?

Ratvar commented:
Can clown PDA come in sans?

TungstenOxide committed:
The build failed so I'm changing it back

Cyberboss reviewed (changes requested):
What @lordpidey said

TungstenOxide committed:
Update PDA.dm

ShizCalev commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -317,13 +317,8 @@ GLOBAL_LIST_EMPTY(PDAs)
-			//font_index = font_index%4
-
+			//CODE REVISION 2
+			font_index = (font_index + 1) % 4
```
wasn't this supposed to be under the if statement?

TungstenOxide committed:
Update PDA.dm

ShizCalev commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -307,6 +316,22 @@ GLOBAL_LIST_EMPTY(PDAs)
+				//CODE REVISION 2
+				font_index = (font_index + 1) % 4
+
+				if (font_index == 0)
```
Use a switch

TungstenOxide replied:
wot

TungstenOxide replied:
Oh hold up.

JJRcop replied:
Switch is
```DM
switch(font_index)
     if(0)
        //VT323
    if(1)
        //Share Tech Mono
```

etc

TungstenOxide replied:
Yup, got it! Did it actually change?

AnturK commented:
These should be on some settings screen not smacked in the middle of every panel.

TungstenOxide committed:
Make the settings visible only on the main screen

mode 0 is the hub, and the hub is the main menu. I think it'd be a lot more convenient for players to always have these options on the main menu if they need them. Makes it really easy to swap colors and fonts at the beginning of the round. I can put it in a settings menu if everyone really thinks it's necessary, but then someone would have to help me actually make the damn icons show up.

TungstenOxide committed:
Update PDA.dm

TungstenOxide committed:
Update PDA.dm

TungstenOxide committed:
Wait, I'm an idiot. I was editing the wrong thing.

JJRcop commented:
Related: HippieStation/HippieStation#3186

TungstenOxide committed:
Here's your switch statements boi

TungstenOxide commented:
@JJRcop yeah, that's my thing. Jamie told me to merge upstream so here I am.

JJRcop commented:
You should download the fonts using the download button, then use our asset cache system, instead of embedding from google
![download_btn](https://user-images.githubusercontent.com/5211576/30294071-dfa914ce-9709-11e7-99ff-b558ee9a5934.PNG)

I did it in #21872 for Font Awesome, used in tgui, use that for reference.

TungstenOxide commented:
@JJRcop Will do, thanks!

JJRcop commented:
This might be a bit of a shortcut: download https://fonts.googleapis.com/css?family=Orbitron|Share+Tech+Mono|VT323

Then download the woff2 files in there, rename them appropriately, put them in an [asset cache](https://github.com/tgstation/tgstation/blob/078fa7c9b13f94356c53df6a66364cc005c95855/code/modules/client/asset_cache.dm) and change the css file to use those new names instead of the url.

TungstenOxide commented:
Leaving future Tungsten a note: DECLARE THE NEW ICONS IN ASSET CACHE

TungstenOxide commented:
I've been wrestling with this for hours and I can't get it to work. @JJRcop can you give me a hand?

JJRcop commented:
Sure, just a moment.

JJRcop commented:
Well a little longer than a moment, doesn't seem to be cooperating with me either.

JJRcop commented:
Okay forget about the asset cache, we can always do it in another PR later, don't worry about it, it works now and I can't get it to work locally so, it's cool.

TungstenOxide reviewed (commented):
K it works

TungstenOxide commented:
@Cyberboss I changed the thing. I think it's ready to merge. I'll figure out how to locally import fonts (eventually) and make a new PR then. 
@Ratvar I'll do the thing later as well.

Cyberboss commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -121,7 +124,8 @@ GLOBAL_LIST_EMPTY(PDAs)
 		return
 
-	var/dat = "<html><head><title>Personal Data Assistant</title></head><body bgcolor=\"#808000\"><style>a, a:link, a:visited, a:active, a:hover { color: #000000; }img {border-style:none;}</style>"
+	var/dat = "<!DOCTYPE html><html><head><title>Personal Data Assistant</title><link href=\"https://fonts.googleapis.com/css?family=Orbitron|Share+Tech+Mono|VT323\" rel=\"stylesheet\"></head><body bgcolor=\"" + background_color + "\"><style>body{" + font_mode + "}ul,ol{list-style-type: none;}a, a:link, a:visited, a:active, a:hover { color: #000000;text-decoration:none; }img {border-style:none;}</st ...
```
Use string embedding with [] instead of +

Cyberboss replied:
dat should be a list that is .Join()'d before browse()

TungstenOxide replied:
That wasn't my decision, there's hundreds, maybe thousands of instances of dat being changed (Since it's the UI of the PDA). Also I haven't a scooby how to do that.

optimumtact replied:
basically

var/datlist = new list()
update all calls to dat +=, to be datlist += "string"

Then at the end of the dat code
var/dat = datlist.Join("") or similar

optimumtact replied:
you only have to do this for the PDA by the way, not every single instance of old dat usage

TungstenOxide replied:
Okay, but why?

Cyberboss commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -307,6 +317,23 @@ GLOBAL_LIST_EMPTY(PDAs)
+						font_mode = "font-family:\"Share Tech Mono\", monospace;letter-spacing:0px;"
+					if (2)
+						font_mode = "font-family:\"Orbitron\", monospace;letter-spacing:0px; font-size:15px"
+					if (3)
```
Make these defines please

TungstenOxide replied:
Defines?

optimumtact replied:
Defines look like this
`#define SS_NO_FIRE 2`
and usually go at the top of the file you are gonna use them in
In your case you would want something like
```
#define FONT_VT  0
#define FONT_MONO 1
#define FONT_ORB 3
#define FONT_OTHER 4
//then in code
switch(font_index)
   if(FONT_VT)
       ... stuff
   if(FONT_MONO)
       ... stuff
... etc
```

makes code more readable, and if someone ever changes the values of each define, it's a lot easier since you only need to change it in one place

Cyberboss commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -307,6 +317,23 @@ GLOBAL_LIST_EMPTY(PDAs)
+					if (3)
+						font_mode = "font-family:monospace;"
+			if ("Change_Color")
+				var/new_color = input("Please enter a color name or hex value (Default is \'#808000\').")
```
why not `as color` instead of hex?

JJRcop replied:
The input proc pulls up the color picker if you add ```as color``` after the closing parenthesis, so you don't have to type in the string or color name. Even if typing "red", or "green" into the field works, the color picker is better.

TungstenOxide replied:
Ohhhhh. Okay lemme do that at some point

Cyberboss commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -26,6 +26,9 @@ GLOBAL_LIST_EMPTY(PDAs)
 	var/obj/item/cartridge/cartridge = null //current cartridge
 	var/mode = 0 //Controls what menu the PDA will display. 0 is hub; the rest are either built in or based on cartridge.
 	var/icon_alert = "pda-r" //Icon to be overlayed for message alerts. Taken from the pda icon file.
+	var/font_index = 0 //The index of the font array that represents the font selected.
```
define

Cyberboss commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -26,6 +26,9 @@ GLOBAL_LIST_EMPTY(PDAs)
 	var/mode = 0 //Controls what menu the PDA will display. 0 is hub; the rest are either built in or based on cartridge.
 	var/icon_alert = "pda-r" //Icon to be overlayed for message alerts. Taken from the pda icon file.
+	var/font_index = 0 //The index of the font array that represents the font selected.
+	var/font_mode = "font-family:\"VT323\", monospace;letter-spacing:1px;" //The currently selected font.
```
This shouldn't be initialized here imo, call set_font_mode() or something in Initialize()

TungstenOxide replied:
The program needs these variables as soon as it starts generating HTML. It's what tells the code what the font is.

TungstenOxide committed:
Declare font_index at the "Toggle_Font" code

I declared it exactly where it's relevant, but I think this looks unclean. I also redefined it which I actually needed to do.

TungstenOxide committed:
Replaced it where it was

cuz the build broke

WJohn commented:
Pretty 👍

TungstenOxide committed:
Make the color changer a picker instead of a field

An attempt

TungstenOxide committed:
Made these defines

@optimumtact @Cyberboss Like this?

optimumtact commented:
those defines look okay to me :+1:

optimumtact commented:
Sorry is this ready?

TungstenOxide commented:
Yes sir, I believe it is.

kevinz000 commented:
oh GOD THIS LOOKS HORRIBLE

Firecage commented:
This....my god. It is so bloody ugly and hurts the eyes so much. And even ignoring how visually unappealing it is, just go to the PDA Messenger menu. Everything is piled right ontop of eachother and makes it quite a strain to search for a specific person.

TungstenOxide commented:
[Mine looks like this](https://imgur.com/a/TYE2J), can you send me a screenshot?
